### PR TITLE
create preview content for tar files

### LINKED ISSUE #372: UFAL/FE-Missing TAR and Video previews
2024/Apr/5 - OK: Show thumbnail in the start of the video preview: https://github.com/ufal/clarin-dspace/pull/1004

2024/Sep/27 - MM: The `<source>` tag is empty.

### PR DESCRIPTION
**Pull request by Paurikova2**
| Phases            | MP | MM  |  MB  | MR |   JM | Total  |
|-----------------|----:|----:|----:|-----:|-----:|-------:|
| ETA                  |  0  |  0  |    0 |     0 |      0 |        0 |
| Developing      |  2  |  0  |    0 |    0 |      0 |         0 |
| Review             |  0  |  0  |    0 |    0 |      0 |         0 |
| Total                |   -  |   -  |   -   |  -    |   -    |         0 |
| ETA est.             |      |      |       |       |         |         0 |
| ETA cust.           |   -  |   -  |   -  |   -   |   -     |        0 |
## Problem description
Show preview for .tar files.

### CODE REVIEW

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/MetadataBitstreamRestRepository.java [hunk 1]
```diff
@@ -390,21 +388,35 @@ public String extractFile(InputStream inputStream, String fileType) {
...
                         }
-                    });
+                    }
+                }
+            } else {
+                zipFileSystem = FileSystems.newFileSystem(tempFile, (ClassLoader) null);
+                Path root = zipFileSystem.getPath("/");
+                Files.walk(root)
+                        .forEach(path -> {
+                            try {
```
Comment (milanmajchrak) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/MetadataBitstreamRestRepository.java:408]: nested try catch  is that right?

**Review by milanmajchrak — CHANGES_REQUESTED** [review]

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/MetadataBitstreamRestRepository.java [hunk 2]
```diff
@@ -390,37 +388,52 @@ public String extractFile(InputStream inputStream, String fileType) {
...
-                        try {
-                            long fileSize = Files.size(path);
-                            if (Files.isDirectory(path)) {
-                                filePaths.add(path.toString().substring(1) + "/|" + fileSize );
-                            } else {
-                                filePaths.add(path.toString().substring(1) + "|" + fileSize );
-                            }
-                        } catch (IOException e) {
-                            e.printStackTrace();
+            if ("tar".equals(fileType)) {
```
Comment (vidiecan) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/MetadataBitstreamRestRepository.java:391]: can't we use less duplicated code? try to copy and paste to github copilot or chatgpt

Comment (vidiecan) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/MetadataBitstreamRestRepository.java:391]: there are many ways how to do it 
- unify parts to one generic function 
- use classes
- ...

**Review by vidiecan — CHANGES_REQUESTED** [review]

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/MetadataBitstreamRestRepository.java (hunk 1 above)
Comment (Paurikova2) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/MetadataBitstreamRestRepository.java:408]: it should be effective for automatic closure of file

**Comment by Paurikova2** [comment]
@milanmajchrak @vidiecan I split the extractFile method into several smaller methods... so many changes because of it

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/MetadataBitstreamRestRepository.java [hunk 3]
```diff
@@ -364,106 +365,153 @@ private String composePreviewURL(Context context, Item item, Bitstream bitstream
...
+     * @param fileType the type of file for which to create a temporary file
+     * @return a Path object representing the temporary file
+     * @throws IOException if an I/O error occurs while creating the file
      */
-    public String extractFile(InputStream inputStream, String fileType) {
-        List<String> filePaths = new ArrayList<>();
-        Path tempFile = null;
-        FileSystem zipFileSystem = null;
+    private Path createTempFile(String fileType) throws IOException {
+        String extension = "tar".equals(fileType) ? ".tar" : ".zip";
```
Comment (milanmajchrak) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/MetadataBitstreamRestRepository.java:375]: use contants for `tar` and `zip`

**Review by milanmajchrak — CHANGES_REQUESTED** [review]

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/MetadataBitstreamRestRepository.java [hunk 4]
```diff
@@ -464,10 +501,46 @@ public String extractFile(InputStream inputStream, String fileType) {
...
+
+        try {
+            // Create a temporary file based on the file type
+            tempFile = createTempFile(fileType);
+
+            // Copy the input stream to the temporary file
+            Files.copy(inputStream, tempFile, StandardCopyOption.REPLACE_EXISTING);
+
+            // Process the file based on its type
+            if ("tar".equals(fileType)) {
```
Comment (milanmajchrak) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/MetadataBitstreamRestRepository.java:528]: What am I seeing? You have defined a constant for this and still using a raw string? nonono

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/MetadataBitstreamRestRepository.java [hunk 5]
```diff
@@ -382,76 +388,107 @@ private String composePreviewURL(Context context, Item item, Bitstream bitstream
...
+    /**
+     * Closes the specified FileSystem resource if it is not null.
+     * @param zipFileSystem the FileSystem to close
+     */
+    private void closeFileSystem(FileSystem zipFileSystem) {
+        if (!Objects.isNull(zipFileSystem)) {
+            try {
+                zipFileSystem.close();
+            } catch (IOException e) {
+                e.printStackTrace();
```
Comment (milanmajchrak) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/MetadataBitstreamRestRepository.java:457]: This e.prinStackTrace does not help anyone. Why we don't use logs?

**Review by milanmajchrak — CHANGES_REQUESTED** [review]

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/MetadataBitstreamRestRepository.java [hunk 6]
```diff
@@ -82,7 +83,10 @@
 @Component(MetadataBitstreamWrapperRest.CATEGORY + "." + MetadataBitstreamWrapperRest.NAME)
 public class MetadataBitstreamRestRepository extends DSpaceRestRepository<MetadataBitstreamWrapperRest, Integer> {
     private static Logger log = org.apache.logging.log4j.LogManager.getLogger(MetadataBitstreamRestRepository.class);
-
+    private final String FILE_EXTENSION_ZIP = ".zip";
+    private final String FILE_EXTENSION_TAR = ".tar";
+    private final String ARCHIVE_TYPE_ZIP = "zip";
```
Comment (vidiecan) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/MetadataBitstreamRestRepository.java:88]: why like this ? isn't it duplicate?

**Review by vidiecan — CHANGES_REQUESTED** [review]

**Review by milanmajchrak — APPROVED** [review]

**Review by vidiecan — APPROVED** [review]

**Pull request merged**